### PR TITLE
fix/update-es6-destructor-challenge: Added test

### LINKED ISSUE #17829: Use Destructuring Assignment to Assign Variables from Nested Objects
#### Describe your problem and - if possible - how to reproduce it

Strict requirement for the solution, requiring only `max: maxOfTomorrow`. 
This became more of an issue when the previous lesson was equally if not more confusing to understand.

Side issue: The console messages are confusing. There are two parts:
1. When there is a syntax error or the likes, Run Tests doesn't work. First reaction was to refresh page and see if things work. So I think a message that there is a syntax error would be helpful. 
2. Error message like `nested destructuring was used` was confusing to understand. My first thought was, is this an error or not when it first appear on the console, because I usually pass in one go. Second thought was, should nested destructuring be used of not. I think better wording should be used like using 'not' to indicate what not to do.

, especially when the previous lesson was equally if not more confusing.

#### Add a Link to the page with the problem
```js

const LOCAL_FORECAST = {
  today: { min: 72, max: 83 },
  tomorrow: { min: 73.3, max: 84.6 }
};

function getMaxOfTmrw(forecast) {
  "use strict";
  // change code below this line
  const {tomorrow: {min, max:maxOfTomorrow}} = forecast; // change this line
  // change code above this line
  return maxOfTomorrow;
}

console.log(getMaxOfTmrw(LOCAL_FORECAST)); // should be 84.6
```
nested destructuring was used

#### Tell us about your browser and operating system
User Agent is: <code>Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.87 Safari/537.36</code>.

**Link to the challenge:**
https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-nested-objects

### LINKED ISSUE #35246: ES6 challenge has poor wording: Use Destructuring Assignment to Assign Variables from Objects
The text of the challenge reads: 

> Use destructuring to obtain the average temperature for tomorrow from the input object AVG_TEMPERATURES, and assign value with key tomorrow to tempOfTomorrow in line.

Except `AVG_TEMPERATURES` is not the input object, `avgTemperatures` the parameter is.  Anyone trying to destructure the global constant will not pass  the challenge.  The simple fix is just to change the one word, from AVG_TEMPERATURES to avgTemperatures.  Alternatively, maybe just pass the object in-line and not use a global constant at all.

### LINKED ISSUE #35520: ES6: destructuring with reassignment was used (Use Destructuring Assignment to Assign Variables from Objects )
**Describe your problem and how to reproduce it:**
In the ES6: destructing assignment to assign variables from object lesson, i make the correct answer but it is not recognized, it shows the test message warning 'destructing with reassignment was used. 
here is the code snippet:
`const AVG_TEMPERATURES = {
  today: 77.5,
  tomorrow: 79
};

function getTempOfTmrw(avgTemperatures) {
  "use strict";
  // change code below this line
  const {tomorrow:tempOfTomorrow} = AVG_TEMPERATURES; // change this line
  // change code above this line
  return tempOfTomorrow;
}

console.log(getTempOfTmrw(AVG_TEMPERATURES)); // should be 79;
;`

**Add a Link to the page with the problem:**
https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-objects/

**Tell us about your browser and operating system:**
* Browser Name: google chrome
* Browser Version: 72.0.3626.119 (Official Build) (64-bit)
* Operating System: windows 10

**If possible, add a screenshot here (you can drag and drop, png, jpg, gif, etc. in this box):**
es6 warning 1

### PR DESCRIPTION
**Pull request by ctwplyp**
- Added test in case during destructuring somebody passes an extra key
- Did not get to test locally

- [ ] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [ ] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [ ] My pull request targets the `master` branch of freeCodeCamp.
- [ ] None of my changes are plagiarized from another source without proper attribution.
- [ ] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [ ] My changes do not use shortened URLs or affiliate links.

Closes #17829 (issue)
Closes #35246 (issue)
Closes #35520 (issue)
Closes #35537 (PR)
Closes #35548 (PR)

### CODE REVIEW

**Comment by thecodingaviator** [comment]
I don't understand the point of adding this test. Why exactly do you need it?

**Comment by ctwplyp** [comment]
Most people will pass the other variable in the destructoring of the
variable and will think it is correct but won't know why it isn't because I
that's what i did when I did the exercise.

**Comment by ojeytonwilliams** [comment]
Let me see if I understand the motivation for this.

The instructions have destructuring assignments of the form 
```js
const { a: one, b: two, c: three, d: four } = anObj;
```
where anObj has all these properties.  However, the challenge wants you to write an expression of the form
```js
const { b: two } = anObj; 
```
which ignores the other three properties.

So, for this challenge it might be tempting to write 

```js
const {today: tempOfToday, tomorrow: tempOfTomorrow } = avgTemperatures;
```

If you do this, the only error you see mentions the lack of destructuring assignment, but you have used it correctly! Just with an extra assignment.

Given all that, I think it's a fine thing to test.  I might word it a little differently.  How about "Today's temperature should be assigned, only tempOfTomorrow."?

Finally, the test itself isn't quite right.  It should be something like `assert.not(...)` since `getUserInput('index')` is just a string and does not have a `.not` function.  I'm a little confused as to how this got through CI checking.

**Comment by ctwplyp** [comment]
Awesome understanding of the problem and great point.  I didn't want to
build the project so this was my best guess at the solution going blind and
not being great at js. Thanks for reviewing! Having you put this much time
into the review makes me much more motivated to contribute too this project
thanks.
  Sean

**Comment by RandellDawson** [comment]
@ctwplyp, @ojeytonwilliams, @thecodingaviator, @chuckadams I think I have come up with a way to resolve the outstanding issues raised on #17829, #35246, and #35520 and which were attempted to be resolved on PRs #35537 and #35548.  Since @ctwplyp's PR first attempted to address part of the issue, I hope it is OK with @ojeytonwilliams and @chuckadams that we fix it on this PR and close the other two.

Instead of changing the instructions, we can make some modifications to what @ctwplyp already started and rearrange the last two tests.  The idea in the 2nd test is to check to see if any destucturing is used.  Then, in the 3rd test, we validate that destructuring for at least the `tomorrow` key is used.

Below you will see my proposed change for all three tests.  I have removed the assert message arguments, as a recent change in the core logic only uses the `text` above each `testString` for this challenge and many other like this one.

By allowing the user to destructure either the global variable `AVG_TEMPERATURES` or the `avgTemperatures` local to the getTempOfTmrw function, we don't have to change the original instructions.  Plus, technically either should work.

```yml
tests:
  - text: <code>getTempOfTmrw(AVG_TEMPERATURES)</code> should be <code>79</code>
    testString: assert(getTempOfTmrw(AVG_TEMPERATURES) === 79);
  - text: destructuring with reassignment was used
    testString: getUserInput => assert(code.match(/{[\S\s]*\w+\s*:[\S\s]*\w+\s*}\s*=\s*(avgTemperatures|AVG_TEMPERATURES)/));
  - text: The key <code>tomorrow</code> was destructured from <code>AVG_TEMPERATURES</code>
    testString: getUserInput => assert(code.match(/{[\S\s]*tomorrow\s*:\s*tempOfTomorrow[\S\s]*}\s*=\s*(avgTemperatures|AVG_TEMPERATURES)/)); 
```
I have tested this locally and tried various edge cases to make sure the regex is flexible yet accurate.  Please let me know if it should be changed or can be simplified further and achieve the same goal.

**Comment by chuckadams** [comment]
Looks good: it's a more robust solution than simply changing the instructions, since it's immune to users stumbling into using the wrong identifier even with the instructions changed.  With this fix, it's not like `getTempOfTmrw` needs to even take a parameter at all, but I'm satisfied enough with the color of this bikeshed now :)

**Comment by ojeytonwilliams** [comment]
@ctwplyp You're very welcome!  It's definitely worth setting up a local copy of the site.  While it's not always strictly necessary (if you're not touching code), it is useful.  It can be slightly fiddly to set up, but the people in the [Gitter room](https://gitter.im/FreeCodeCamp/Contributors) will help you out.

@RandellDawson I checked it locally and it accepts all the kinds of solutions we've been talking about and should cut down the confusion.  Looks good.

**Comment by RandellDawson** [comment]
@ctwplyp Please go ahead and change all the tests to what I show above and then other reviewers will review/merge this PR.

Thank you.

**Comment by kneekoo** [comment]
This covers a lot of cases with all kinds of spacing and variable names for "today" and it even allows missing the ending semicolon, which is syntactically correct:
```
\s*const\s*{\s*(today\s*(:\s*[a-z]+)?\s*,)?\s*tomorrow\s*:\s*tempOfTomorrow\s*}\s*=\s*avgTemperatures\s*;?
```

Validated samples:
```
const { today:tempOfToday, tomorrow: tempOfTomorrow}=avgTemperatures;
const { today:tempToday, tomorrow: tempOfTomorrow}=avgTemperatures;
const{today,tomorrow:tempOfTomorrow}=avgTemperatures;
const { today, tomorrow: tempOfTomorrow } = avgTemperatures
const { tomorrow : tempOfTomorrow } = avgTemperatures;
```

Now if you want to allow the use of AVG_TEMPERATURES in the const declaration, I would recommend triggering a warning so the users know that the point of passing input to a function's argument is to actually use that argument, not an outside const/variable. Either this, or don't accept using the global const. Because it doesn't make sense to disallow destructuring today but accept bypassing the function's argument.

**Comment by RandellDawson** [comment]
> Because it doesn't make sense to disallow destructuring today but accept bypassing the function's argument.

Global constants are still accessible via a function.

**Comment by kneekoo** [comment]
> the regex suggestion I made above works for all the cases @kneekoo pointed out,

It does work, but it's wildly permissive with all kinds of wrong code, because [\S\s]* is wrong in this case. Take a look at these:

**good code** validated
```
const { today:tempOfToday, tomorrow: tempOfTomorrow}=avgTemperatures;
const { today:tempToday, tomorrow: tempOfTomorrow}=avgTemperatures;
const{today,tomorrow:tempOfTomorrow}=avgTemperatures;
const { today, tomorrow: tempOfTomorrow } = avgTemperatures
const { tomorrow : tempOfTomorrow } = avgTemperatures;
```

**bad code** validated because of [\S\s]*
```
const { today:, tomorrow: tempOfTomorrow}=avgTemperatures;
const { :tempToday, tomorrow: tempOfTomorrow}=avgTemperatures;
const{,tomorrow:tempOfTomorrow}=avgTemperatures;
const { today, : tempOfTomorrow } = avgTemperatures
const { tomorrow tempOfTomorrow } = avgTemperatures;
```

And as I said, it makes no sense to disallow destructuring "today" while allowing the use or the global constant. But if both are allowed, it should still be hinted that the point of the function is to use its argument to provide the correct result, and that "today" is unnecessary to provide the correct result and should be left out of the destructuring declaration.

If we want to allow the global constant, this is it:
```
\s*const\s*{\s*(today\s*(:\s*\w+)?\s*,)?\s*tomorrow\s*:\s*tempOfTomorrow\s*}\s*=\s*(avgTemperatures|AVG_TEMPERATURES)\s*;?
```

If the global constant shouldn't be allowed, then my first regex applies properly validates the good code and invalidates the bad code:
```
\s*const\s*{\s*(today\s*(:\s*\w+)?\s*,)?\s*tomorrow\s*:\s*tempOfTomorrow\s*}\s*=\s*avgTemperatures\s*;?
```

**Comment by RandellDawson** [comment]
@kneekoo Your regex is still too restrictive, as yours would not allow:
```js
const { tomorrow: tempOfTomorrow, today}=avgTemperatures;
```
Also, every single one of your examples of invalid (bad) syntax would cause all the tests to fail anyone, so we don't have to worry about those cases with my regex.  My regex even allows extra variables like (not that they are needed) for this challenge.  You have to test it locally to see how the various code examples will pass/fail.  You can not just rely on if the regex captures or does not capture something.  That is why I was able to use [\S\s]* and not worry about the invalid syntax that this would potentially allow. 
```js
const { regexIsFun, tomorrow: tempOfTomorrow } = avgTemperatures;
```
If you test my suggestion locally you will see what I am talking about.

There may be some strange edge cases not handled by my regex that have not even been discovered yet, but it handles all that I have ever seen complained about on the forum or on GitHub issues.

**Comment by kneekoo** [comment]
@RandellDawson My regex can surely be expanded to allow a mixed order of keys, but I see no reason to accept syntactically invalid code by relying on the browser's JS engine parser. And the fact that your regex covers all the strange cases you've encountered is only a result of using regex that is more permissive than it should.

And this is also wrong and shouldn't be allowed because **regexIsFun** is not a key in the passed object:
```
const { regexIsFun, tomorrow: tempOfTomorrow } = avgTemperatures;
```

It would be great if all the cases would be gathered here so we can test the proposed regexes against them. Do you have the links at hand, by any chance?

Updated regex allowing AVG_TEMPERATURES (although I would at least warn against using it):
```
\s*const\s*{\s*(today\s*(:\s*\w+)?\s*,\s*)?tomorrow\s*:\s*tempOfTomorrow\s*(,\s*today\s*(:\s*\w+)?\s*)?}\s*=\s*(avgTemperatures|AVG_TEMPERATURES)\s*;?
```

**Comment by lasjorg** [comment]
I see no point in even having a function, especially not if we allow the user to do destructuring of the global. This seems to be a common occurrence with the challenges, adding extra complexity where it is not needed.

Why not just have the object and destructuring with no function:

```js
const AVG_TEMPERATURES = {
  today: 77.5,
  tomorrow: 79
};

const tempOfTomorrow = AVG_TEMPERATURES.tomorrow; // change this line to use destructuring instead
```

Instead of voxel and some random decimal numbers, why not something more relatable for the example code:

```js
var user = {name: 'John Doe', age: 34, password: 'qwerty' };

var name = user.name; // name = 'John Doe'
var age = user.age; // age = 34
var password = user.password; // password = 'qwerty'

const { name, age, password } = user; // name = 'John Doe', age = 34, password = 'qwerty'

const { name: tempName, age: tempAge, password: tempPassword } = user; // tempName = 'John Doe', tempAge = 34, tempPassword = 'qwerty'
```

I can't seem to come up with anything better for the instructions right now, it is still not great.

Instructions:

> Use destructuring to get the value of the tomorrow property from the object AVG_TEMPERATURES. Assign the value to a new variable named tempOfTomorrow in line.

**Comment by RandellDawson** [comment]
@lasjorg If you want to suggest completely changing the challenge, please create an issue for discussion first.  This PR is about improving the challenge as it is currently.  What has been suggested will resolve the original issue.

**Comment by kneekoo** [comment]
> I see no point in even having a function, especially not if we allow the user to do destructuring of the global. This seems to be a common occurrence with the challenges, adding extra complexity where it is not needed.

At this stage in the JS course, using a function shouldn't pose any kind of problem to someone who got this far. If anyone has issues with functions, their place is obviously not in the ES6 section of the course.

**Comment by seanlaing** [comment]
Update the challenge with your commented changes please review @kneekoo  @RandellDawson @lasjorg @chuckadams .

**Comment by RandellDawson** [comment]
@ctwplyp You were only suppose to add the 3 tests that had in my suggestion.  You have added a fourth which should not have been added.  Just copy/paste what I showed in my last comment to you above.

@lasjorg Added his suggestions to an issue which is being discussed and @kneekoo's suggestion is incorrect.

@kneekoo The following is not "wrong".  It is completely valid syntax and will not cause an error of any kind. 
```
const { regexIsFun, tomorrow: tempOfTomorrow } = avgTemperatures;
```

We don't care if the user tries to add other non-existent object properties.  All we care about in this challenge is that the user properly destructures `tomorrow` from the object and name it `tempOfTomorrow`.  The regex I suggested already handles invalid code, because it will not pass any of the tests.  If you want to debate what we should allow, please create a separate issue so it can be discussed.  

If @ctwplyp can make the changes I suggested, we can get this PR merged much faster, because it fixes 3 issues.

**Comment by kneekoo** [comment]
> We don't care if the user tries to add other non-existent object properties. All we care about in this challenge is that the user properly destructures `tomorrow` from the object and name it `tempOfTomorrow`.

In the real world we want the developers to write good code, considering software should do everything we expect it to, and nothing else. I don't know if you speak for everyone at freeCodeCamp but not caring about bad code is like not caring about vulnerabilities just because the code does what it's supposed to do.

You probably know who [Robert C. Martin](https://en.wikipedia.org/wiki/Robert_C._Martin) is, so how about we help people to develop healthy coding habits by helping them early on to write good and clean code?

**Comment by RandellDawson** [comment]
@kneekoo If we wanted to technical here about clean code, we would keep the original tests in place and not allow the extra `today` property as it is not needed or used.  But that is not the point of this exercise.  The point is to properly destructure tomorrow from the object and rename it to something else.  

The tests I proposed validate the user has properly done this and at the same time remains flexible enough for the user to add other properties (which exist or not in the object).  

I am all for teaching the user to write good and clean code, but we also want to remain flexible enough to allow valid code to still pass.  This challenge confused many campers, because the original tests were too restrictive.  Now they will be much less restrictive and still test the destructuring objects concept.

There is no point in discussing this any further on this PR.  If you want to change the entire scope of the challenge, please create an issue and it can be discussed there.

Thank you.

**Comment by kneekoo** [comment]
@RandellDawson Even in the current state, the exercise can be improved in the provided documentation, the requirements and the performed tests. It's strange seeing how after I gave some feedback on how these aspects can be improved you kept nitpicking at my proposals.

You should've noticed how I responded constructively to your observations by adjusting the regex to fit a more varied set of solutions, but whether you noticed or not, you kept dismissing everything even when I showed you how your suggestion allows for syntactically invalid code. And all these when we can offer the users better feedback and make sure that their code is actually good. Learning proper coding from the beginning is important because it creates better developers and it saves a lot of time with relearning.

So please don't lose sight of the fact that this is an educational project and while you might be tempted to dismiss the feedback that you don't care about, the purpose of freeCodeCamp must favor the students. They need better documentation, better feedback on tested code, and test suits that help them write sane code that will benefit everyone once they take it in the real world.

And for this to happen, you should be more welcoming with the constructive feedback that you get. Because virtually no one will know/care who wrote some test, but if it helps a lot of people, then it's great.

**Comment by RandellDawson** [comment]
@kneekoo It is not that I don't agree with many of your points.  It is that they need to be discussed on separate issue which would be focused on improvement many aspects of the challenge.  Since October, we (volunteers) have reviewed 10K+ PRs and still have another 3K+ PRs to review.  We can not make these "perfect".  If a PR makes an incremental improvement, then we are grateful for that.  We could spend months debating on this PR about what the ultimate regex should be, but at the end of the day, we want to get the changes in which improve them.

We try not to allow any changes to the tests which would allow the user to pass the challenge with incorrect code.  None of the changes I suggested to do.  Do they allow some code which may not be considered "best" practice by some?  Very possibly.  That can be discussed in the separate issue and then as a community, we can decide what we want to teach.  All this PR is trying to do at this point is resolve three outstanding issues and remain as flexible as possible.

I am closing this to further comments except other mods/members, so they can make their comments once @ctwplyp makes the changes suggested by me which are the most flexible and still fix the problems addressed in the 3 issues.

Look forward to you creating your issue where we can discuss your other suggestions further.

Thank you.

**Comment by RandellDawson** [comment]
I made the changes to the tests on behalf of the user in an effort to get this PR merged faster.  I also reverted the change to the `AVG_TEMPERATURE` in a PR which should not have been merged.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-objects.english.md [hunk 1]
```diff
@@ -19,7 +19,7 @@ You may read it as "get the field <code>x</code> and copy the value into <code>a
 
 ## Instructions
 <section id='instructions'>
-Use destructuring to obtain the average temperature for tomorrow from the input object <code>avgTemperatures</code>, and assign value with key <code>tomorrow</code> to <code>tempOfTomorrow</code> in line.
+Use destructuring to obtain the average temperature for tomorrow from the input object <code>AVG_TEMPERATURES</code>, and assign value with key <code>tomorrow</code> to <code>tempOfTomorrow</code> in line.
```
Comment (moT01) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-objects.english.md]:
```suggestion
Use destructuring to obtain the average temperature for tomorrow from the input object <code>avgTemperatures</code>, and assign value with key <code>tomorrow</code> to <code>tempOfTomorrow</code> in line.
```
This was changes to `avgTemperatures` 5 days ago - I think it should remain like that. If we aren't going to destructure the passed in parameter what's the point of having a parameter.

**Review by moT01 — CHANGES_REQUESTED** [review]
After quite some time spent reading this book of an issue and trying to understand the related issues... I think this looks good enough if we wanted to merge it. But, I do think this one thing should be changed...

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-objects.english.md (hunk 1 above)
Comment (RandellDawson) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-objects.english.md]: @moT01 It was merged, but it was not supposed to be.  It had been agreed by even the OP of the other PR, that it was fine to leave it as AVG_TEMPERATURES.  There is nothing wrong with the original instruction of using AVG_TEMPERATURES.  It is completely valid JavaScript to use the globallly declared logic or the parameter inside the function.

Comment (RandellDawson) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-objects.english.md]: Also, as it was already stated to @lasjorg in a previous comment, if we want to completely change the challenge code and move away from a function, then that discussion should take place in a completely separate issue, so unless there is invalid JavaScript here, please merge.

Comment (moT01) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/es6/use-destructuring-assignment-to-assign-variables-from-objects.english.md]: Sounds good

**Review by moT01 — APPROVED** [review]

**Review by moT01 — APPROVED** [review]
Looks good @ctwplyp, congratulations on your first contribution to freeCodeCamp! 🎊 There's plenty more to do if you're interested. In the future, be sure to complete the checklist in the PR template. Thanks and happy coding!
